env-builder/src/flox-env.cc

```diff
@@ -0,0 +1,309 @@
+    {
+      throw Error( "No packages found for system '%s'", system );
+    }
+
+  /* extract all packages */
+
+  std::vector<resolver::LockedPackageRaw> locked_packages;
+
+  for ( auto const & package : packages->second )
+    {
+      if ( ! package.second.has_value() ) { continue; }
+      auto const & locked_package = package.second.value();
+      locked_packages.push_back( locked_package );
+    }
+
+  /**
+   * extract derivations
+   */
+  StorePathSet                      references;
+  std::vector<StorePathWithOutputs> drvsToBuild;
+  Packages                          pkgs;
+
+  for ( auto const & package : locked_packages )
+    {
+
+      auto package_input_ref = FlakeRef( package.input );
+      auto package_flake
+        = flake::lockFlake( state, package_input_ref, flake::LockFlags {} );
+
+      auto vFlake = state.allocValue();
+      flake::callFlake( state, package_flake, *vFlake );
+      state.forceAttrs( *vFlake, noPos, "while parsing flake" );
+
+
+      auto output = vFlake->attrs->get( state.symbols.create( "outputs" ) );
+
+
+      /* evaluate the package */
+      for ( auto path_segment : package.attrPath )
+        {
+          state.forceAttrs( *output->value,
+                            output->pos,
+                            "while parsing cached flake data" );
+
+          auto next
+            = output->value->attrs->get( state.symbols.create( path_segment ) );
+
+          if ( ! next )
+            {
+              std::ostringstream str;
+              output->value->print( state.symbols, str );
+              throw Error( "Attribute '%s' not found in set '%s'",
+                           path_segment,
+                           str.str() );
+            }
+          output = next;
+        }
+
+
+      auto package_drv = getDerivation( state, *output->value, false );
+
+      if ( ! package_drv.has_value() )
+        {
+          throw Error( "Failed to get derivation for package '%s'",
+                       nlohmann::json( package ).dump().c_str() );
+        }
+
+      /* Collect all outputs to include in the environment */
+      for ( auto output : package_drv->queryOutputs() )
+        {
+          if ( ! output.second.has_value() )
+            {
+              continue;
+            }  // skip outputs without path
+          pkgs.emplace_back(
+            state.store->printStorePath( output.second.value() ),
+            true,
+            package.priority );
+          references.insert( output.second.value() );
+        }
+
+      /* Collect drvs that may yet need to be built */
+      if ( auto drvPath = package_drv->queryDrvPath() )
+        {
+          drvsToBuild.push_back( { *drvPath } );
+        }
+    }
+
+  /* Build derivations that make up the environment */
+  // todo check if this builds `outputsToInstall` only
+  state.store->buildPaths( toDerivedPaths( drvsToBuild ),
+                           state.repair ? bmRepair : bmNormal );
+
+  // todo: is it script _xor_ file?
```

Comment:
Can we expand this comment with the current assumption how things should work. And bring it up for discussion.